### PR TITLE
Switch to Bundler gem structure

### PR DESCRIPTION
**Pull request by citizen428**
Just moving with the times and turning this into a contemporary gem.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -138,19 +170,17 @@ Or you can toggle it inside IRB/Pry:
...
-I initially wrote this for the students of the core Ruby course on
-[RubyLearning](http://rubylearning.org), so Rails is not of interest to me (not
-saying it doesn't work there, just that I test in plain IRB/Pry and not with
-the Rails console.
+## Contributing
 
-## Thanks
+Bug reports and pull requests are welcome on GitHub at https://github.com/citizen428/methodfinder.
+
+### Noteworthy contributors
```
Comment (chocolateboy) [inline README.md]: I can't speak for the other contributors, but I think a changelog would be better. Thanks/credits could be included there manually, or automatically via a changelog generator (I recommend [this one](https://github.com/github-changelog-generator/github-changelog-generator) ([alternatives](https://github.com/github-changelog-generator/github-changelog-generator/wiki/Alternatives))).

It would also be a good idea to add git tags for each release (that tool uses them). I might add them if you don't beat me to it :-)

Comment (chocolateboy) [reply README.md]: [Done](https://github.com/citizen428/methodfinder/releases).

These are mostly based on when the version changed in the [`.gemspec`](https://github.com/citizen428/methodfinder/commits/master/methodfinder.gemspec). If you know better, please update the tags. You'll need to delete and re-push each one e.g.:

```
$ git tag -d v1.2.3
$ git push --delete origin v1.2.3
$ git tag -a v1.2.3 -m v1.2.3 <commit>
$ git push --tags
```

File: README.md [hunk 2]
```diff
@@ -138,19 +170,17 @@ Or you can toggle it inside IRB/Pry:
...
 
 * [Matthew Lucas](https://github.com/lucas-matt) for [first packaging this as a gem](https://github.com/citizen428/methodfinder/pull/1).
 * [Ryan Bates](https://github.com/ryanb) for [suggesting](https://github.com/citizen428/methodfinder/issues/closed#issue/3)
```
Comment (chocolateboy) [inline README.md]: Broken links:

* https://github.com/citizen428/methodfinder/issues/closed#issue/3 -> https://github.com/citizen428/methodfinder/issues/3
* https://github.com/citizen428/methodfinder/issues/closed#issue/4 -> https://github.com/citizen428/methodfinder/issues/4

File: .gitignore [hunk 3]
```diff
@@ -1,5 +1,8 @@
-pkg/
-*.rbc
-.bundle/
-doc/
-/vendor/
```
[the remark this answers is not in the document]
Comment (citizen428) [reply .gitignore]: Sure, fine. I never use a `vendor/` directory, but if you do ¯\_(ツ)_/¯

File: Gemfile [hunk 4]
```diff
@@ -1,13 +1,6 @@
 source 'https://rubygems.org'
 
-group :test do
-  gem 'minitest'
-end
+git_source(:github) { |repo_name| "https://github.com/#{repo_name}" }
```
[the remark this answers is not in the document]
Comment (citizen428) [reply Gemfile]: This file is auto-generates by bundler and everyone familiar with Ruby should have seen it plenty of times by now. I added it anyway.

File: README.md [hunk 5]
```diff
@@ -6,29 +6,61 @@
...
+
+Common sense not included!
+
+While I never had any problems with this, it's still better to be safe than
+sorry, so use this with caution and maybe not on production data.
+
+I initially wrote this for the students of the core Ruby course on
+[RubyLearning](http://rubylearning.org), so Rails is not of interest to me (not
+saying it doesn't work there, just that I test in plain IRB/Pry and not with
+the Rails console.
```
[the remark this answers is not in the document]
Comment (citizen428) [reply README.md]: This must have been missing for a long time, thanks.

File: Rakefile [hunk 6]
```diff
@@ -1,14 +1,8 @@
 # frozen_string_literal: true
 
-require 'rake/testtask'
+require 'bundler/gem_tasks'
 require 'rdoc/task'
```
[the remark this answers is not in the document]
Comment (citizen428) [reply Rakefile]: Yes, it is needed, because it's configuring Rdoc to only include specific files, otherwise `LICENSE.md`, `Gemfile` etc. needlessly show up in the generated docs.

Which version of Ruby are you using, `rdoc/task` should be in the standard library: https://docs.ruby-lang.org/en/2.5.0/RDocTask.html

File: methodfinder.gemspec [hunk 7]
```diff
@@ -1,12 +1,25 @@
...
+  spec.version       = MethodFinder::VERSION
+  spec.authors       = ['Michael Kohl']
+  spec.email         = ['citizen428@gmail.com']
+
+  spec.summary       = 'A Smalltalk-like Method Finder for Ruby'
+  spec.description   = 'A Smalltalk-like Method Finder for Ruby with some extra features'
+  spec.homepage      = 'http://citizen428.github.com/methodfinder/'
+  spec.license       = 'MIT'
+
+  spec.files         = `git ls-files -z`.split("\x0").reject do |f|
```
[the remark this answers is not in the document]
Comment (citizen428) [reply methodfinder.gemspec]: Bundler autogenerated, changed.

File: README.md (hunk 2 above)
Comment (citizen428) [reply README.md]: Removed the whole section because of #19.

File: README.md [hunk 8]
```diff
@@ -165,22 +195,5 @@ the Rails console.
...
-copies or substantial portions of the Software.
-
-THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
-IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
-FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
-AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
-LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
-OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
-SOFTWARE.
+The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).
```
[the remark this answers is not in the document]
Comment (citizen428) [reply README.md]: I figure this link + the LICENSE file makes it less necessary to include the whole license text here again.

File: .travis.yml [hunk 9]
```diff
@@ -13,3 +13,4 @@ before_install:
 matrix:
   allow_failures:
       - rvm: rbx-3
+
```
[the remark this answers is not in the document]
Comment (citizen428) [reply .travis.yml]: Fixed, thanks.

File: methodfinder.gemspec [hunk 10]
```diff
@@ -1,12 +1,25 @@
...
+
+Gem::Specification.new do |spec|
+  spec.name          = 'methodfinder'
+  spec.version       = MethodFinder::VERSION
+  spec.authors       = ['Michael Kohl']
+  spec.email         = ['citizen428@gmail.com']
+
+  spec.summary       = 'A Smalltalk-like Method Finder for Ruby'
+  spec.description   = 'A Smalltalk-like Method Finder for Ruby with some extra features'
+  spec.homepage      = 'http://citizen428.github.com/methodfinder/'
```
[the remark this answers is not in the document]
Comment (citizen428) [reply methodfinder.gemspec]: I had set all the metadata manually on Rubygems.org, but this is better, agreed.

You can take care of adding the `changelog` metadata in #19 if you want.

File: Rakefile (hunk 6 above)
[the remark this answers is not in the document]
Comment (chocolateboy) [reply Rakefile]: ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-linux]

**Comment by chocolateboy** [comment]
> I'm tempted to move JRuby to the allow_failures where Rubinius already is. Thoughts?

Yes. That's fine by me.

File: Rakefile (hunk 6 above)
[the remark this answers is not in the document]
Comment (citizen428) [reply Rakefile]: Hm, `ruby 2.5.1p57 (2018-03-29 revision 63029) [x86_64-darwin17]` here, so same just on a Mac. 

Does `ruby -e 'require "rdoc/task"'` raise an exception for you?

**Comment by citizen428** [comment]
Ok, JRuby is now allowed to fail.

File: Rakefile (hunk 6 above)
[the remark this answers is not in the document]
Comment (chocolateboy) [reply Rakefile]: Thanks for your help. With your prompting, I think I've resolved this :tada:

TL;DR:

Please add the following to `methodfinder.gemspec`:

```ruby
spec.add_development_dependency 'rdoc', '~> 6.0'
```

Longer TL;DR:

* on my (Arch Linux) system, `rdoc/task` is provided by a [packaged gem](https://www.archlinux.org/packages/community/any/ruby-rdoc/), not the standard library
* `bundler exec` runs commands in a locked-down environment (`$LOAD_PATH`) which only includes standard library files and explicitly-installed project dependencies
* because `rdoc/task` is neither in the standard library (on my system) nor (currently) provided by a dependency, it is not picked up
* adding the [rdoc gem](https://rubygems.org/gems/rdoc) as a development dependency solves this problem on my system (and shouldn't hurt on systems which already have `rdoc/task.rb` in the standard library)

---

> Does `ruby -e 'require "rdoc/task"'` raise an exception for you?

No, that works fine for me outside of Bundler. However:

```
$ bundler exec ruby -e 'require "rdoc/task"'

Traceback (most recent call last):
	1: from -e:1:in `<main>'
-e:1:in `require': cannot load such file -- rdoc/task (LoadError)
```

Note: on my (Arch Linux) system, the RDoc task file resolves to `/usr/lib/ruby/gems/2.5.0/gems/rdoc-6.0.3/lib/rdoc/task.rb`, which is not a core ruby file (the [ruby](https://www.archlinux.org/packages/extra/x86_64/ruby/) package only provides `/usr/lib/ruby/2.5.0/rdoc.rb`). It's provided by an optional package, [ruby-rdoc](https://www.archlinux.org/packages/community/any/ruby-rdoc/), which I (recently) installed manually. If I uninstall that package, the error occurs outside of Bundler:

```
$ ruby -e 'require "rdoc/task"'

Traceback (most recent call last):
	2: from -e:1:in `<main>'
	1: from /usr/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
/usr/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require': cannot load such file -- rdoc/task (LoadError)
```

I tried to rectify this yesterday by adding rdoc as a dev dependency, but Bundler appeared to ignore/skip its installation. But I've just tried it now and it appears to work:

```
$ bundler add rdoc

Fetching rdoc 6.0.3
Installing rdoc 6.0.3

$ bundler exec rake

# Running:
...
11 runs, 27 assertions, 0 failures, 0 errors, 0 skips
```

[the remark this answers is not in the document]
Comment (chocolateboy) [reply Rakefile]: Footnote (in case anyone else encounters this issue):

> I tried to rectify this yesterday by adding rdoc as a dev dependency, but Bundler appeared to ignore/skip its installation.

This command:

```
$ bundler add rdoc
```

\- ignores/doesn't install rdoc for me on the (pre-bundlerized) master branch:

```
Fetching gem metadata from https://rubygems.org/
Using rake 12.3.1
Using bundler 1.16.1
Using minitest 5.11.3
```

This appears to be because it's already specified as a (Rubinius) dependency.

```ruby
group :development do
  gem 'rdoc', platform: :rbx
end
```

It works if those lines are commented out:

```
$ bundler add rdoc

Fetching gem metadata from https://rubygems.org/
Resolving dependencies...
Using rake 12.3.1
Using bundler 1.16.1
Using minitest 5.11.3
Fetching rdoc 6.0.3
Installing rdoc 6.0.3
```

**Comment by chocolateboy** [comment]
Incidentally, I'm not seeing any changes on the bundlerize branch yet. Have they been pushed?

**Comment by citizen428** [comment]
They have been now. Once the Travis build finishes I will merge this branch.

File: methodfinder.gemspec (hunk 7 above)
[the remark this answers is not in the document]
Comment (chocolateboy) [reply methodfinder.gemspec]: This doesn't appear to have been pushed. I can add it to (before) #19.
